### PR TITLE
JWT claims

### PR DESCRIPTION
**Pull request by Joannis**
This PR adds all of the standard claims to JWT.

### CODE REVIEW

**Comment by vzsg** [comment]
Is the PEM parsing feature still work-in-progress? Because it's unfortunately not as simple as parsing Base64 between the header and the footer... The binary data inside is a DER-encoded structure, from which you have to extract the actual key data.

**Comment by Joannis** [comment]
That shouldn't have been pushed to this branch

File: Sources/JWT/JWS.swift [hunk 1]
```diff
@@ -18,6 +18,26 @@ enum JWTError: Error {
...
+    
+    case claimExpired
+    
+    case claimedTooSoon
+    
+    case invalidAudience
+}
+
+public struct Signer<C: JWT> {
+    public typealias Verifier = ((C) throws -> Void)
```
Comment (tanner0101) [inline Sources/JWT/JWS.swift]: can we typealias `((C) throws -> Void)`

File: Sources/JWT/JWS.swift [hunk 2]
```diff
@@ -18,6 +18,26 @@ enum JWTError: Error {
...
+    
+    var secret: Data
+    var identifier: String
+    
+    public var verifications = [Verifier]()
```
Comment (tanner0101) [inline Sources/JWT/JWS.swift]: `public var verifications: [Verifier]`

File: Sources/JWT/JWS.swift [hunk 3]
```diff
@@ -166,6 +186,32 @@ public struct JSONWebSignature<C: Codable>: Codable {
         default:
             throw JWTError.invalidJWS
         }
+        
+        try self.verify(claim: payload)
+    }
+    
+    fileprivate func verify(claim: C) throws {
+        if let claim = claim as? ExpirationClaim {
```
Comment (tanner0101) [inline Sources/JWT/JWS.swift]: casting is slow. we should have the method for verifying on the claim itself and just call that.

File: Sources/JWT/JWT.swift [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import Foundation
+
+/// A JWT is a Publically Readable set of claims
+///
+/// Each variable represents a claim
+public protocol JWT: Codable {}
+
+/// Identifies by which application a JWT is issued
+public protocol IssuerClaim: JWT {
```
Comment (tanner0101) [inline Sources/JWT/JWT.swift]: An `IssuerClaim` is a `Claim`, not an entire `JWT`.

File: Sources/JWT/JWT.swift [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import Foundation
+
+/// A JWT is a Publically Readable set of claims
+///
+/// Each variable represents a claim
+public protocol JWT: Codable {}
```
Comment (tanner0101) [inline Sources/JWT/JWT.swift]: I think there's more to a `JWT` than just being codable. You should have an array of claims that are `get/set` as well as methods for validating it

File: Sources/JWT/JWT.swift [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+/// Identifies the date at which a JWT was issued
+public protocol IssuedAtClaim: JWT {
+    /// The `Date` at which this claim was issued
+    var iat: Date { get }
+}
+
+/// Identifies the date at which a JWT was issued
+public protocol IDClaim: JWT {
```
Comment (tanner0101) [inline Sources/JWT/JWT.swift]: These protocol seems more like `IDClaimContaining` or `IDClaimable`. The actual claim itself should be a struct.

File: Tests/JWTTests/JWSTests.swift [hunk 7]
```diff
@@ -39,11 +45,97 @@ class JWSTests: XCTestCase {
...
     static var allTests: [(String, (JWSTests) -> () throws -> Void)] = [
-        ("testBasics", testBasics)
+        ("testBasics", testBasics),
+        ("testTimeClaims", testTimeClaims),
+        ("testAudienceClaim", testAudienceClaim),
     ]
 }
 
-struct AuthenticationMessage : Codable {
+struct TimeClaim: JWT, ExpirationClaim, NotBeforeClaim, IssuedAtClaim {
```
Comment (tanner0101) [inline Tests/JWTTests/JWSTests.swift]: It seems like `JWT` is being used here more like `JWTPayload`. The `JWT` class should be concrete, generic on a supplied JWTPayload type.

File: Sources/JWT/JWS.swift (hunk 1 above)
Comment (Joannis) [reply Sources/JWT/JWS.swift]: This already Is a type alias, right?

File: Sources/JWT/JWT.swift [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    func verify() throws
+}
+
+/// A claim is a codable top-level property that can be verified against the current circumstances
+///
+/// Multiple claims form a payload
+public protocol Claim: Codable {}
+
+/// Identifies by which application a Claim is issued
+public protocol IssuerClaim: Claim {
```
Comment (tanner0101) [inline Sources/JWT/JWT.swift]: I don't think the usability of this is great. It makes more sense to me for claims to be structs that conform to `Claim`.

```swift
protocol Claim: Codable {
    func verify() throws
}
```

```swift
struct ExpirationClaim: Claim {
    var expirationDate: Date
    func verify() throws { ... }
}
```

```swift
protocol JWTPayload: Codable {
    var claims: [Claim] { get }
}
```

```swift
import JWT 

struct MyPayload: JWTPayload {
    var exp: ExpirationClaim
    var foo: String

    var claims: [Claim] { return [exp] }
}
```

### Or with key paths

```swift
protocol JWTPayload: Codable {
   static  var claims: [KeyPath<Self, Claim>] { get }
}
```

```swift
import JWT 

struct MyPayload: JWTPayload {
    var exp: ExpirationClaim
    var foo: String

    static let claims = [\.exp]
}
```

Comment (Joannis) [reply Sources/JWT/JWT.swift]: I thought we discussed this the other way around. That all properties should be inside the token as an actual variable and that they should be validated.

**Pull request merged**